lib/Parse/ParsePattern.cpp

```diff
@@ -783,40 +784,18 @@ Parser::parseFunctionSignature(Identifier SimpleName,
   Status |= parseFunctionArguments(NamePieces, bodyParams, paramContext,
                                    defaultArgs);
   FullName = DeclName(Context, SimpleName, NamePieces);
-  
-  // Check for the 'throws' keyword.
-  rethrows = false;
-  if (Tok.is(tok::kw_throws)) {
-    throwsLoc = consumeToken();
-  } else if (Tok.is(tok::kw_rethrows)) {
-    throwsLoc = consumeToken();
-    rethrows = true;
-  } else if (Tok.isAny(tok::kw_throw, tok::kw_try)) {
-    throwsLoc = consumeToken();
-    diagnose(throwsLoc, diag::throw_in_function_type)
-      .fixItReplace(throwsLoc, "throws");
-  }
-
-  SourceLoc arrowLoc;
 
-  auto diagnoseInvalidThrows = [&]() -> Optional<InFlightDiagnostic> {
-    if (throwsLoc.isValid())
-      return None;
-
-    if (Tok.is(tok::kw_throws)) {
-      throwsLoc = consumeToken();
-    } else if (Tok.is(tok::kw_rethrows)) {
-      throwsLoc = consumeToken();
-      rethrows = true;
-    }
+  // Check for the 'async' keyword.
+  if (Context.LangOpts.EnableExperimentalConcurrency &&
+      Tok.isContextualKeyword("async")) {
+    asyncLoc = consumeToken();
+  }
 
-    if (!throwsLoc.isValid())
-      return None;
+  // Check for the 'async' and 'throws' keywords.
+  rethrows = false;
+  parseAsyncThrows(SourceLoc(), asyncLoc, throwsLoc, &rethrows);
```

Comment:
788-792 is a leftover? If it's really needed, could you clarify why it can't be parsed in `parseAsyncThrows`?

Reply:
Definitely a leftover, thank you!